src/main/java/at/hannibal2/skyhanni/data/PetAPI.kt

```diff
@@ -35,31 +178,492 @@ object PetAPI {
+            cleanName = "",
+            level = 0,
+            rarity = rarity,
+            petItem = petInfo.heldItem?.asInternalName(),
+            xp = petInfo.exp,
+            rawPetName = "",
+        )
+    }
+
+    private fun parsePetName(displayName: String): PetData? {
+        petNameMenuPattern.matchMatcher(displayName) {
+            val name = group("name").orEmpty()
+            val rarity = LorenzRarity.getByColorCode(group("rarity")[0]) ?: ErrorManager.skyHanniError(
+                "Couldn't parse pet rarity.",
+                Pair("displayName", displayName),
+                Pair("rarity", group("rarity"))
+            )
+            val level = group("level").toInt()
+            val skin = group("skin").orEmpty()
+
+            return PetData(
+                internalName = petNameToInternalName(name, rarity),
+                cleanName = name,
+                rarity = rarity,
+                petItem = null,
+                level = level,
+                xp = 0.0,
+                rawPetName = skin,
+            )
+        }
+        return null
+    }
+
+    @HandleEvent
+    fun onCommandRegistration(event: CommandRegistrationEvent) {
+        event.register("shpetxp") {
+            description = "Calculates the pet xp from a given level and rarity."
+            category = CommandCategory.DEVELOPER_TEST
+            callback { levelToXPCommand(it) }
+        }
+    }
+
+    private fun levelToXPCommand(input: Array<String>) {
+        if (input.size < 3) {
+            ChatUtils.userError("Usage: /shcalcpetxp <level> <rarity> <pet>")
+            return
+        }
+
+        val level = input[0].toIntOrNull()
+        if (level == null) {
+            ChatUtils.userError("Invalid level '${input[0]}'.")
+            return
+        }
+        val rarity = LorenzRarity.getByName(input[1])
+        if (rarity == null) {
+            ChatUtils.userError("Invalid rarity '${input[1]}'.")
+            return
+        }
+
+        val petName = input.slice(2..<input.size).joinToString(" ")
+        val xp: Double = levelToXP(level, rarity, petName) ?: run {
+            ChatUtils.userError("Invalid level or rarity.")
+            return
+        }
+        ChatUtils.chat(xp.addSeparators())
+        return
+    }
+
+    private fun levelToXP(level: Int, rarity: LorenzRarity, petName: String): Double? {
+        val newPetName = petNameToFakeInternalName(petName)
+
+        val rarityOffset = getRarityOffset(rarity, newPetName) ?: return null
+        if (!isValidLevel(level, newPetName)) return null
+
+        val xpList = xpLeveling + getCustomLeveling(newPetName)
+
+        return xpList.slice(0 + rarityOffset..<level + rarityOffset - 1).sum().toDouble()
+    }
+
+    private fun isValidLevel(level: Int, petName: String): Boolean {
+        val petsData = petsDataNEU ?: run {
+            ErrorManager.skyHanniError("NEUPetsData is null")
+        }
+
+        val maxLevel = petsData.get(petName)?.maxLevel ?: 100
```

Comment:
Replace function call with indexed accessor.

Reply:
dont use 100 as default, return true or false insetead (more clear the behaviour)